### PR TITLE
Fix tf.math.segment_max/min/mean/sun/prod crashes(aborts) when segment_ids is large

### LINKED ISSUE #46888: tf.math.segment_max/min/mean/sun/prod crashes(aborts) when segment_ids is large
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below):2.1.0
- Python version:3.7.6
- Bazel version (if compiling from source):N/A
- GCC/Compiler version (if compiling from source):N/A
- CUDA/cuDNN version:N/A
- GPU model and memory:N/A

**Describe the current behavior**
tf.math.segment_max/min/mean/sun/prod crashes(aborts) when `segment_ids` is large

**Describe the expected behavior**
expect an exception message if the input is unexpected instead of crash

**Standalone code to reproduce the issue**
~~~python
tf.math.segment_max(data=np.ones((1,10,1)), segment_ids=[1676240524292489355])
tf.math.segment_min(data=np.ones((1,10,1)), segment_ids=[1676240524292489355])
tf.math.segment_mean(data=np.ones((1,10,1)), segment_ids=[1676240524292489355])
tf.math.segment_sum(data=np.ones((1,10,1)), segment_ids=[1676240524292489355])
tf.math.segment_prod(data=np.ones((1,10,1)), segment_ids=[1676240524292489355])
~~~

Output:
~~~python
2021-02-03 16:44:25.849065: F tensorflow/core/framework/tensor_shape.cc:405] Check failed: 0 <= new_num_elements (0 vs. -1684338830784658056)
Aborted (core dumped)
~~~

### PR DESCRIPTION
**Pull request by yongtang**
This PR fixes the issue raised in #46888 where tf.math.segment_max/min/mean/sun/prod crashes(aborts) when segment_ids is large.
    
This PR fixes #46888.
    
Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Review by mihaimaruseac — APPROVED** [review]

File: tensorflow/core/kernels/segment_reduction_ops_impl.h [hunk 1]
```diff
@@ -110,7 +110,7 @@ class SegmentReductionOp : public OpKernel {
                 errors::InvalidArgument("Shape must be at least rank 1"));
 
     TensorShape output_shape = input.shape();
-    output_shape.set_dim(0, output_rows);
+    OP_REQUIRES_OK(context, output_shape.SetDimWithStatus(0, output_rows));
```
Comment (sanjoy) [inline tensorflow/core/kernels/segment_reduction_ops_impl.h:113]: Why doesn't the 

```
    OP_REQUIRES(context, output_rows >= 0,
                errors::InvalidArgument("segment ids must be >= 0"));
```

check above catch this?

File: tensorflow/core/kernels/segment_reduction_ops_impl.h [hunk 2]
```diff
@@ -250,7 +250,7 @@ class SegmentReductionGPUOp : public AsyncOpKernel {
 
     if (num_indices == 0) {
       TensorShape output_shape = input.shape();
-      output_shape.set_dim(0, 0);
+      OP_REQUIRES_OK_ASYNC(context, output_shape.SetDimWithStatus(0, 0), done);
```
Comment (sanjoy) [inline tensorflow/core/kernels/segment_reduction_ops_impl.h:253]: What error is this catching?

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/segment_reduction_ops_impl.h (hunk 1 above)
Comment (mihaimaruseac) [reply tensorflow/core/kernels/segment_reduction_ops_impl.h:113]: `SetDim` also checks that the total tensor size does not overflow.

**Review by mihaimaruseac — APPROVED** [review]

File: tensorflow/core/kernels/segment_reduction_ops_impl.h (hunk 1 above)
Comment (sanjoy) [reply tensorflow/core/kernels/segment_reduction_ops_impl.h:113]: Thanks, maybe add a comment?

**Review by mihaimaruseac — APPROVED** [review]

**Pull request merged**